source/fab/steps/cleanup_prebuilds.py

```diff
@@ -0,0 +1,179 @@
+# ##############################################################################
+#  (c) Crown copyright Met Office. All rights reserved.
+#  For further details please refer to the file COPYRIGHT
+#  which you should have received as part of this distribution
+# ##############################################################################
+"""
+Pruning of old files from the incremental/prebuild folder.
+
+"""
+import logging
+import os
+import time
+from datetime import timedelta, datetime
+from pathlib import Path
+from time import sleep
+from typing import Dict, Optional, Iterable, Set
+
+from fab.constants import CURRENT_PREBUILDS
+from fab.steps import Step
+from fab.util import file_walk, get_fab_workspace, get_prebuild_file_groups
+
+logger = logging.getLogger(__name__)
+
+
+class CleanupPrebuilds(Step):
+    """
+    A step to delete old files from the local incremental/prebuild folder.
+
+    Assumes prebuild filenames follow the pattern: `<stem>.<hash>.<suffix>`.
+
+    """
+    def __init__(self, older_than: Optional[timedelta] = None, n_versions: int = 0, all_unused: Optional[bool] = None):
```

Comment:
Can we come up with a better API? At the moment it's possible to specify none, one or more options and we don't have logic to handle all those cases.

Does Python support multiple method definition polymorphism? i.e. `def __init__(self)`, `def __init__(self, older_than: timedelta)`, `def __init__(self, n_versions: int)`, etc.

Or is the goal to support multiple modes of operation, e.g. cleanup anything older than 5 days plus anything which hasn't been used?

Reply:
The goal is to support both modes like you say, but also I believe we _do_ handle all the cases you mention.
 - If we provide nothing, then `all_unused` is set
 - If we provide nothing for the first two, then `all_unused` _must_ be set.
 - One or both of the first two is fine.

The only thing we might want to do it make clear that when `all_unused` is set there's no point providing the first two because they are ignored. Perhaps that should raise an error?